### ISSUE TITLE
no-member complaint when using `__new__` set member of the class

### ISSUE THREAD

**Issue by tonyxwz** [PyCQA/pylint#3158]
When it is needed to write abstract class for user to further extend, I usually use the `__new__` magic method to define some members of the class so that users can have access without having to call the `__init__` method. But pylint seems to complain on this.

Is it a bug or it is not recommended to write in this way?

### Steps to reproduce
```python
""" vscode 1.38.1
"""
from abc import ABC, ABCMeta, abstractmethod

class FooBase(ABC):
    def __new__(cls, *args, **kwargs):
        instance = super().__new__(cls)
        instance.name = 'Foo'
        return instance

    def __init__(self):
        self.name2 = 'Foo2'

    @abstractmethod
    def myrepr(self):
        return self.name # <-- here complains pylint(no-member)

class Foo(FooBase):
    def __init__(self, name=""):
        # super().__init__()
        self.name = name

    def myrepr(self):
        basename = super().myrepr()
        return str({'base': basename, 'name': self.name})


if __name__ == "__main__":
    f = Foo(name='bar')
    print(f.myrepr())
```

### Current behavior
pylint(no-member)

### Expected behavior
no complaints 

### pylint --version output
pylint 2.3.1
astroid 2.2.5
Python 3.6.9 |Anaconda, Inc.| (default, Jul 30 2019, 14:00:49) [MSC v.1915 64 bit (AMD64)]

**Comment by PCManticore** [comment]
Thanks for opening an issue.

There's no way for `pylint` to know that `name` is going to exist at runtime as we do not analyze `__new__` constructors. In the future you could use typed annotations to indicate that certain members are going to exist at runtime as in:

```
class FooBase(...):
   name: int

class Foo(metaclass=FooBase)
    ...

Foo().name # won't trigger no-member any longer
```

**Comment by zhangzhiqiangcs** [comment]
> Thanks for opening an issue.
> 
> There's no way for `pylint` to know that `name` is going to exist at runtime as we do not analyze `__new__` constructors. In the future you could use typed annotations to indicate that certain members are going to exist at runtime as in:
> 
> ```
> class FooBase(...):
>    name: int
> 
> class Foo(metaclass=FooBase)
>     ...
> 
> Foo().name # won't trigger no-member any longer
> ```

Sorry to bother you. 
This is a sample code,
```
class A(BaseModel):
  query: Query

class B(A):
  pass

print(B.query)
```
There is query member absolutely. Using python run the code, no error. but pylint give a no-member query error.
